### ISSUE TITLE
How to initialize the `enc_with_bottleneck.vq` as paper said?

### ISSUE THREAD

**Issue by Ruiyuan-Zhang** [lucidrains/discrete-key-value-bottleneck-pytorch#1]
Hi, 

1. when i use the next code snippet to load a pretain model as paper's setting, the output dim is `(batch_size, 2048)`, so how can i input it to `enc_with_bottleneck`?
```python 
resnet50 = torch.hub.load('facebookresearch/dino:main', 'dino_resnet50')
```
2. just as the above setting, how can i use that pertrained model to initialize my `enc_with_bottleneck.vq`? especially how to initialize the `codebook's keys`

3. i try to slove my problem with the next code snippet, but my evluation on testset show accuracy is 0.1(just like random choice)
```python
from discrete_key_value_bottleneck_pytorch import DiscreteKeyValueBottleneck
enc_with_bottleneck = DiscreteKeyValueBottleneck(
    dim = 384,             
    num_memory_codebooks=64, 
    num_memories = 128,     
    dim_memory = 32,     
    decay = 0.8,            # the exponential moving average decay, lower means the keys will change faster
)
enc_with_bottleneck.to(device)
optimizer_kvib = optim.SGD(enc_with_bottleneck.parameters(), lr=0.001,momentum=0.8)


for x in enc_with_bottleneck.vq.parameters(): 
    x.requires_grad = True
optimizer_kvib_vq = optim.SGD(enc_with_bottleneck.vq.parameters(), lr=0.001,momentum=0.8)
for e in range(5):
    for i, data in tqdm(enumerate(trainloader, 0)):
        inputs, labels = data
        inputs, labels = inputs.to(device), labels.to(device)

        x = resnet50(inputs)
        print(x.shape)
        x = torch.unsqueeze(x,dim=1)
        optimizer_kvib_vq.zero_grad()
        _, _, loss = enc_with_bottleneck.vq(x)
        loss.backward()
        optimizer_kvib_vq.step()

for x in enc_with_bottleneck.vq.parameters(): 
    x.requires_grad = False
```

I look forward to your reply
zhang ruiyuan

**Comment by pedrocg42** [comment]
Hi Zhang, 

I hope you had already solved this question, anyway in this [repo](https://github.com/pedrocg42/discrete-key-value-bottleneck/blob/main/networks/discrete_key_value_bottleneck.py) I was able to replicate their results. More specifically you can find [here](https://github.com/pedrocg42/discrete-key-value-bottleneck/blob/main/networks/discrete_key_value_bottleneck.py) how to create the model with both DKVB and VectorQuantizer.

I will show the way I have been able to replicate their results in a simple snippet:
```python
class DKVBClassifier(nn.Module):
    def __init__(
        self,
        embedding_dim: int = 1024,
        codebook_size: int = 8192,
        num_codebooks: int = 1,
        vq_decay: float = 0.99,
        threshold_ema_dead_code: int = 1e-4,
        value_dimension: Union[int, str] = "same",
    ):

        self.embedding_dim = embedding_dim
        self.codebook_size = codebook_size
        self.num_codebooks = num_codebooks
        self.value_dimension = value_dimension
        self.vq_decay = vq_decay
        self.threshold_ema_dead_code = threshold_ema_dead_code

        # embedding dimension and number of key-value pairs must be divisible by number of codes
        assert (self.embedding_dim % num_codebooks) == 0
        assert (self.codebook_size & num_codebooks) == 0

        # Getting pretrained encoder with DINO
        encoder_raw = torch.hub.load("facebookresearch/dino:main", "dino_resnet50")

        # Getting all the model up to the second to last bottleneck block, as the paper says. (output dim 1024 as in the paper)
        self.encoder = nn.Sequential(
            encoder_raw.conv1,
            encoder_raw.bn1,
            encoder_raw.relu,
            encoder_raw.maxpool,
            encoder_raw.layer1,
            encoder_raw.layer2,
            encoder_raw.layer3,
        )

        # Freezing the encoder
        for param in self.encoder.parameters():
            param.requires_grad = False

        self.dkvb= DiscreteKeyValueBottleneck(
            dim=self.embedding_dim,  # input dimension
            codebook_dim=self.embedding_dim // self.num_codebooks,
            num_memory_codebooks=self.num_codebooks,  # number of memory codebook
            num_memories=self.codebook_size,  # number of memories
            dim_memory=self.embedding_dim
            // self.num_codebooks,  # dimension of the output memories
            decay=self.vq_decay,  # the exponential moving average decay, lower means the keys will change faster
            threshold_ema_dead_code=self.threshold_ema_dead_code,  # (0.1·batch-size·h·w/num-pairs)
        )

        self.decoder = nn.Sequential(
            nn.Linear(self.embedding_dim, 512),
            nn.ReLU(),
            nn.Linear(self.embedding_dim, 256),
            nn.ReLU(),
            nn.Linear(self.embedding_dim, 10),
            nn.Softmax(dim=1),
        )

    def forward(self, x):

        with torch.no_grad():
            embeddings = self.encoder(x)
            embeddings.detach_()


        encoder_output_size = embeddings.shape[-1]
        batch_size = x.size()[0]

        embeddings = torch.reshape(
            embeddings,
            (embeddings.shape[0], self.embedding_dim, encoder_output_size**2),
        )  # B, Dim, H, W -> B, Dim, N
        embeddings = torch.permute(embeddings, (0, 2, 1))  # B, Dim, N -> B, N, Dim

        memories = self.dkvb(embeddings)

        memories = torch.permute(memories, (0, 2, 1))  # B, N, Dim -> B, Dim, N
        memories = torch.reshape(
            memories,
            (
                batch_size,
                self.embedding_dim,
                encoder_output_size,
                encoder_output_size,
            ),
        )  # B, Dim, N -> B, Dim, H, W

        # Processing final output
        return self.decoder(memories)
```

To initialize your keys before your training process you only have to leave the dkvb in train mode and do the inference for N epochs, being N=10 in the paper. You do not need gradient descent for the keys to converge so doing this should be fine.

```python
model = DKVBClassifier()
model.to(device)

model.eval()
model.dkvb.train()
with torch.no_grad():
    for epoch in range(10):
        for images, labels in dataset:
            images.to(device)
            model(images)
```

I hope this helps, please feel free to reach out to me in case you need more help! :)

**Comment by Ruiyuan-Zhang** [comment]
Wow, you are so nice !!!
Thanks for your code and a really clear explanation. 

![image](https://user-images.githubusercontent.com/71813586/230618198-313a9bd9-1e5a-4560-a849-43f4bfba8299.png)
